lightning/src/ln/channel.rs

```diff
@@ -4016,11 +4016,23 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
 				_ => {}
 			}
 		}
+		let funding_txo = if let Some(funding_txo) = self.funding_txo {
+			// If we haven't yet exchanged funding signatures (ie channel_state < FundingSent),
+			// returning a channel monitor update here would imply a channel monitor update before
+			// we even registered the channel monitor to begin with, which is invalid.
+			// Thus, if we aren't actually at a point where we could conceivably broadcast the
+			// funding transaction, don't return a funding txo (which prevents providing the
+			// monitor update to the user, even if we return one).
+			// See test_duplicate_chan_id and test_pre_lockin_no_chan_closed_update for more.
+			if self.channel_state & (ChannelState::FundingSent as u32 | ChannelState::ChannelFunded as u32 | ChannelState::ShutdownComplete as u32) != 0 {
```

Comment:
By chance would `if self.channel_state >= ChannelState::FundingSent as u32 {` work instead? passes tests and a bit more readable...

Reply:
I don't think so - I think that would match `FundingCreated|MonitorUpdateFailed` which would be wrong. `channel_state` is a bit confusing with the bit flags.

Reply:
I think `FundingCreated` < `FundingSent`? But true, I guess technically we can have monitor updates starting from the `FundingCreated` state onward, oh well.

Reply:
Right, so we'd think `FundingCreated|MonitorUpdateFailed` > `FundingSent` because the second flag has a higher bit position.

Reply:
Oh duh, my brain read `|` as the word "or"